File: src/main/java/io/github/aalsanie/codes/Issue.java
package io.github.aalsanie.codes;

import java.util.Objects;
import org.jspecify.annotations.Nullable;

public final class Issue {
    private final @Nullable OutcomeCode code;
    private final @Nullable String path;
    private final String message;

    private Issue(@Nullable OutcomeCode code, @Nullable String path, String message) {
        this.code = code;
        this.path = path;
        this.message = message;
    }

    public static Issue of(String message) {
        return create(null, null, message);
    }

    public static Issue coded(OutcomeCode code, String message) {
        return create(Objects.requireNonNull(code, "code"), null, message);
    }

    public static Issue at(String path, String message) {
        return create(null, Objects.requireNonNull(path, "path"), message);
    }

    public static Issue at(String path, OutcomeCode code, String message) {
        return create(
            Objects.requireNonNull(code, "code"),
            Objects.requireNonNull(path, "path"),
            message
        );
    }

    public @Nullable OutcomeCode getCode() {
        return code;
    }

    public @Nullable String getPath() {
        return path;
    }

    public String getMessage() {
        return message;
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) {
            return true;
        }
        if (!(other instanceof Issue issue)) {
            return false;
        }
        return Objects.equals(code, issue.code)
            && Objects.equals(path, issue.path)
            && message.equals(issue.message);
    }

    @Override
    public int hashCode() {
        return Objects.hash(code, path, message);
    }

    @Override
    public String toString() {
        StringBuilder builder = new StringBuilder();
        if (code != null) {
            builder.append(code).append(' ');
        }
        if (path != null) {
            builder.append(path).append(": ");
        }
        return builder.append(message).toString();
    }

    private static Issue create(
        @Nullable OutcomeCode code,
        @Nullable String path,
        String message
    ) {
        if (path != null) {
            Constraints.requirePath(path);
        }
        Constraints.requireMessage(message);
        return new Issue(code, path, message);
    }
}

File: src/main/kotlin/io/github/aalsanie/codes/Issue.kt
package io.github.aalsanie.codes

import io.github.aalsanie.codes.internal.Constraints

public class Issue private constructor(
    public val code: OutcomeCode?,
    public val path: String?,
    public val message: String,
) {
    public override fun equals(other: Any?): Boolean =
        this === other ||
            (
                other is Issue &&
                    code == other.code &&
                    path == other.path &&
                    message == other.message
            )

    public override fun hashCode(): Int {
        var result = code?.hashCode() ?: 0
        result = 31 * result + (path?.hashCode() ?: 0)
        result = 31 * result + message.hashCode()
        return result
    }

    public override fun toString(): String =
        buildString {
            if (code != null) {
                append(code)
                append(' ')
            }
            if (path != null) {
                append(path)
                append(": ")
            }
            append(message)
        }

    public companion object {
        @JvmStatic
        public fun of(message: String): Issue = create(null, null, message)

        @JvmStatic
        public fun coded(
            code: OutcomeCode,
            message: String,
        ): Issue = create(code, null, message)

        @JvmStatic
        public fun at(
            path: String,
            message: String,
        ): Issue = create(null, path, message)

        @JvmStatic
        public fun at(
            path: String,
            code: OutcomeCode,
            message: String,
        ): Issue = create(code, path, message)

        private fun create(
            code: OutcomeCode?,
            path: String?,
            message: String,
        ): Issue {
            if (path != null) {
                Constraints.requirePath(path)
            }
            Constraints.requireMessage(message)
            return Issue(code, path, message)
        }
    }
}
